### PR TITLE
Add support for URL params to server endpoint.

### PR DESCRIPTION
**Pull request by Raynos**
This PR adds support URL params for the server.

There are a few subtle nuances. If there are required thrift fields
that come from URL params you need to mark them as "not in JSON"

Currently this only supports reading strings from the URL params
and not booleans/floats/ints.

r: @uber/zanzibar-team

### CODE REVIEW

File: codegen/test_data/clients/bar.gogen [hunk 1]
```diff
@@ -532,7 +532,7 @@ func (c *barClient) ArgWithParams(
 	req := zanzibar.NewClientHTTPRequest(c.clientID, "ArgWithParams", c.httpClient)
 
 	// Generate full URL.
-	fullURL := c.httpClient.BaseURL + "/bar" + "/" + string(r.UUID) + "/segment" + "/" + string(r.Params.UserUUID)
+	fullURL := c.httpClient.BaseURL + "/bar" + "/argWithParams" + "/" + string(r.UUID) + "/segment" + "/" + string(r.Params.UserUUID)
```
Comment (ravirajj) [inline codegen/test_data/clients/bar.gogen]: Is `fmt.Sprintf()` faster than string concatenation?

File: examples/example-gateway/idl/endpoints/bar/bar.thrift [hunk 2]
```diff
@@ -35,7 +35,10 @@ struct QueryParamsOptsStruct {
 }
 
 struct ParamsStruct {
-    1: required string userUUID (zanzibar.http.ref = "params.user-uuid")
+    1: required string userUUID (
+        zanzibar.http.ref = "params.user-uuid"
+        go.tag = "json:\"-\""
```
Comment (ravirajj) [inline examples/example-gateway/idl/endpoints/bar/bar.thrift]: Does this remove the field from the json body?

Comment (Raynos) [reply examples/example-gateway/idl/endpoints/bar/bar.thrift]: yes it does.

File: codegen/test_data/clients/bar.gogen (hunk 1 above)
Comment (Raynos) [reply codegen/test_data/clients/bar.gogen]: not sure.

File: codegen/method.go [hunk 3]
```diff
@@ -431,10 +443,89 @@ func findParamsAnnotation(
 	walkFieldGroups(fields, visitor)
 
 	if identifier == "" {
-		return "", false
+		return "", required, false
```
Comment (cl1337) [inline codegen/method.go]: nit: more than one return of the same type may want to label all of them, i.e.   (required, ok bool)

File: codegen/test_data/clients/bar.gogen (hunk 1 above)
Comment (cl1337) [reply codegen/test_data/clients/bar.gogen]: fmt.Sprintf allocates a new string object and it's slow, use https://golang.org/pkg/bytes/#Buffer.WriteString if performance is a concern, which seems not here

File: examples/example-gateway/idl/endpoints/contacts/contacts.thrift [hunk 4]
```diff
@@ -30,7 +30,7 @@ struct Contact {
 }
 
 struct SaveContactsRequest {
-    1: required UUID userUUID (zanzibar.http.ref = "params.userUUID");
+    1: required string userUUID (zanzibar.http.ref = "params.userUUID");
```
Comment (cl1337) [inline examples/example-gateway/idl/endpoints/contacts/contacts.thrift]: curious why

Comment (Raynos) [reply examples/example-gateway/idl/endpoints/contacts/contacts.thrift]: if it's uuid then we have to generate code like `args.Request.userUUID = UUID(req.Params.ByName('userUUID'))`

Generating that convert a string to `UUID` alias logic is non-trivial.

**Pull request merged**